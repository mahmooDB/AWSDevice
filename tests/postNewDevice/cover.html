
<!DOCTYPE html>
<html>
	<head>
		<meta http-equiv="Content-Type" content="text/html; charset=utf-8">
		<style>
			body {
				background: black;
				color: rgb(80, 80, 80);
			}
			body, pre, #legend span {
				font-family: Menlo, monospace;
				font-weight: bold;
			}
			#topbar {
				background: black;
				position: fixed;
				top: 0; left: 0; right: 0;
				height: 42px;
				border-bottom: 1px solid rgb(80, 80, 80);
			}
			#content {
				margin-top: 50px;
			}
			#nav, #legend {
				float: left;
				margin-left: 10px;
			}
			#legend {
				margin-top: 12px;
			}
			#nav {
				margin-top: 10px;
			}
			#legend span {
				margin: 0 5px;
			}
			.cov0 { color: rgb(192, 0, 0) }
.cov1 { color: rgb(128, 128, 128) }
.cov2 { color: rgb(116, 140, 131) }
.cov3 { color: rgb(104, 152, 134) }
.cov4 { color: rgb(92, 164, 137) }
.cov5 { color: rgb(80, 176, 140) }
.cov6 { color: rgb(68, 188, 143) }
.cov7 { color: rgb(56, 200, 146) }
.cov8 { color: rgb(44, 212, 149) }
.cov9 { color: rgb(32, 224, 152) }
.cov10 { color: rgb(20, 236, 155) }

		</style>
	</head>
	<body>
		<div id="topbar">
			<div id="nav">
				<select id="files">
				
				<option value="file0">device-db/tests/postNewDevice/main.go (89.6%)</option>
				
				</select>
			</div>
			<div id="legend">
				<span>not tracked</span>
			
				<span class="cov0">not covered</span>
				<span class="cov8">covered</span>
			
			</div>
		</div>
		<div id="content">
		
		<pre class="file" id="file0" style="display: none">// postNewDevice
package main

import (
        "encoding/json"
        "errors"
        "os"

        "data"

        "github.com/aws/aws-lambda-go/events"
        "github.com/aws/aws-lambda-go/lambda"
        "github.com/aws/aws-sdk-go/aws"
        "github.com/aws/aws-sdk-go/aws/session"
        "github.com/aws/aws-sdk-go/service/dynamodb"
        "github.com/aws/aws-sdk-go/service/dynamodb/dynamodbattribute"
)

// Global AWS session variables
var sess *session.Session
var errSess error

// Init function creates an globally accessible AWS session
func init() <span class="cov8" title="1">{
        // Initialize an aws session
        region := os.Getenv("AWS_REGION")
        sess, errSess = session.NewSession(&amp;aws.Config{
                Region: &amp;region},
        )
}</span>

// createNewDevice: takes a string mapped json and checks for requested fields
// if all of required fields are provided returns a new Device object
// Input: a string mapped json
// Output: a new Device object
func createNewDevice(jsonMap map[string]interface{}) (data.Device, error) <span class="cov8" title="1">{

        // Check if input fields are missing
        errFlag := false
        errMsg := "Bad request\nMissing following field(s):"
        if jsonMap["id"] == nil </span><span class="cov8" title="1">{
                errMsg = errMsg + "\n id"
                errFlag = true
        }</span>
        <span class="cov8" title="1">if jsonMap["deviceModel"] == nil </span><span class="cov8" title="1">{
                errMsg = errMsg + "\n deviceModel"
                errFlag = true
        }</span>
        <span class="cov8" title="1">if jsonMap["name"] == nil </span><span class="cov8" title="1">{
                errMsg = errMsg + "\n name"
                errFlag = true
        }</span>
        <span class="cov8" title="1">if jsonMap["note"] == nil </span><span class="cov8" title="1">{
                errMsg = errMsg + "\n note"
                errFlag = true
        }</span>
        <span class="cov8" title="1">if jsonMap["serial"] == nil </span><span class="cov8" title="1">{
                errMsg = errMsg + "\n serial"
                errFlag = true
        }</span>
        // If any fields are missing, return error
        <span class="cov8" title="1">errMsg = errMsg + "\n"
        if errFlag == true </span><span class="cov8" title="1">{
                return data.Device{}, errors.New(errMsg)
        }</span>

        // Create new device object otherwise
        <span class="cov8" title="1">newDevice := data.Device{Id: jsonMap["id"].(string),
                DeviceModel: jsonMap["deviceModel"].(string),
                Name:        jsonMap["name"].(string),
                Note:        jsonMap["note"].(string),
                Serial:      jsonMap["serial"].(string),
        }

        return newDevice, nil</span>
}

// Handler: responsible for taking POST requests from user that provide a json
// containing new Device object and store it in DynamoDB table
// Input: json containing new device information
// Output: json echo of newly inserted device
func Handler(request events.APIGatewayProxyRequest) (events.APIGatewayProxyResponse, error) <span class="cov8" title="1">{

        // Try to unmarshal input json to a map
        var mappedJson map[string]interface{}
        err := json.Unmarshal([]byte(request.Body), &amp;mappedJson)
        // If the input json is not a valid json, return error 500
        if err != nil </span><span class="cov8" title="1">{
                return events.APIGatewayProxyResponse{
                        Body:       "Internal Server Error 1\nJson unmarshaling error: " + err.Error(),
                        StatusCode: 500,
                }, nil
        }</span>

        // Try to create a new Device struct form input
        <span class="cov8" title="1">newDevice, err := createNewDevice(mappedJson)
        // If the input was not compliant with predefined struct, return error 400
        if err != nil </span><span class="cov8" title="1">{
                return events.APIGatewayProxyResponse{
                        Body:       err.Error(),
                        StatusCode: 400,
                }, nil
        }</span>

        // If somthing went wrong with session creation, return error 500
        <span class="cov8" title="1">if errSess != nil </span><span class="cov0" title="0">{
                return events.APIGatewayProxyResponse{
                        Body:       "Internal Server Error 2\nSession error: " + errSess.Error(),
                        StatusCode: 500,
                }, nil
        }</span>
        // Create DynamoDB client
        //db := dynamodb.New(sess)
        <span class="cov8" title="1">svc := dynamodb.New(sess)
        // Converting to mock DynamoDB client for test
        db := MockDynamoDB{svc}

        // Try to prepare a DynamoDB item structure from Device
        dbDevice, err := dynamodbattribute.MarshalMap(newDevice)
        // If somthing went wrong during data prepration, return error 500
        if err != nil </span><span class="cov0" title="0">{
                return events.APIGatewayProxyResponse{
                        Body:       "Internal Server Error 3\nDatabase unmarshaling error: " + err.Error(),
                        StatusCode: 500,
                }, nil
        }</span>

        // Get table name from OS
        <span class="cov8" title="1">tableName := aws.String(os.Getenv("DEVICES_TABLE_NAME"))
        // Prepare a structure for putting the new item
        dbInput := &amp;dynamodb.PutItemInput{
                Item:      dbDevice,
                TableName: tableName,
        }
        // Try to put new item into DynamoDB table
        _, err = db.PutItem(dbInput)
        // If somthing went wrong with database, return error 500
        if err != nil </span><span class="cov0" title="0">{
                return events.APIGatewayProxyResponse{
                        Body:       "Internal Server Error 4\nDatabase error: " + err.Error(),
                        StatusCode: 500,
                }, nil
        }</span>

        // Create a json response from newly inserted device
        <span class="cov8" title="1">jsonResponse, err := json.Marshal(newDevice)
        // If somthing went wrong with json creation, return error 500
        if err != nil </span><span class="cov0" title="0">{
                return events.APIGatewayProxyResponse{
                        Body:       "Internal Server Error 5\nJson marshaling error: " + err.Error(),
                        StatusCode: 500,
                }, nil
        }</span>

        // Finally, everything went smoothly! Echo new device information.
        <span class="cov8" title="1">return events.APIGatewayProxyResponse{
                Body:       string(jsonResponse),
                StatusCode: 201,
        }, nil</span>

}

func main() <span class="cov0" title="0">{
        lambda.Start(Handler)
}</span>
</pre>
		
		</div>
	</body>
	<script>
	(function() {
		var files = document.getElementById('files');
		var visible;
		files.addEventListener('change', onChange, false);
		function select(part) {
			if (visible)
				visible.style.display = 'none';
			visible = document.getElementById(part);
			if (!visible)
				return;
			files.value = part;
			visible.style.display = 'block';
			location.hash = part;
		}
		function onChange() {
			select(files.value);
			window.scrollTo(0, 0);
		}
		if (location.hash != "") {
			select(location.hash.substr(1));
		}
		if (!visible) {
			select("file0");
		}
	})();
	</script>
</html>
