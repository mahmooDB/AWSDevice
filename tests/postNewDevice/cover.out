mode: set
device-db/tests/postNewDevice/main.go:23.75,28.26 3 1
device-db/tests/postNewDevice/main.go:32.2,32.35 1 1
device-db/tests/postNewDevice/main.go:36.2,36.28 1 1
device-db/tests/postNewDevice/main.go:40.2,40.28 1 1
device-db/tests/postNewDevice/main.go:44.2,44.30 1 1
device-db/tests/postNewDevice/main.go:49.2,50.21 2 1
device-db/tests/postNewDevice/main.go:55.2,62.23 2 1
device-db/tests/postNewDevice/main.go:28.26,31.3 2 1
device-db/tests/postNewDevice/main.go:32.35,35.3 2 1
device-db/tests/postNewDevice/main.go:36.28,39.3 2 1
device-db/tests/postNewDevice/main.go:40.28,43.3 2 1
device-db/tests/postNewDevice/main.go:44.30,47.3 2 1
device-db/tests/postNewDevice/main.go:50.21,52.3 1 1
device-db/tests/postNewDevice/main.go:69.93,75.16 3 1
device-db/tests/postNewDevice/main.go:83.2,85.16 2 1
device-db/tests/postNewDevice/main.go:93.2,98.16 3 1
device-db/tests/postNewDevice/main.go:106.2,113.16 4 1
device-db/tests/postNewDevice/main.go:121.2,130.16 4 1
device-db/tests/postNewDevice/main.go:138.2,140.16 2 1
device-db/tests/postNewDevice/main.go:148.2,151.8 1 1
device-db/tests/postNewDevice/main.go:75.16,80.3 1 1
device-db/tests/postNewDevice/main.go:85.16,90.3 1 1
device-db/tests/postNewDevice/main.go:98.16,103.3 1 0
device-db/tests/postNewDevice/main.go:113.16,118.3 1 0
device-db/tests/postNewDevice/main.go:130.16,135.3 1 0
device-db/tests/postNewDevice/main.go:140.16,145.3 1 0
device-db/tests/postNewDevice/main.go:155.13,157.2 1 0
